cohttp-eio/src/reader.ml

```diff
@@ -283,40 +20,40 @@ let token =
 
 let ows = skip_while (function ' ' | '\t' -> true | _ -> false)
 let crlf = string "\r\n"
-let is_cr = function '\r' -> true | _ -> false
+let not_cr = function '\r' -> false | _ -> true
 let space = char '\x20'
-let p_meth = token <* space >>| Http.Method.of_string
+let p_meth = let+ meth = token <* space in Http.Method.of_string meth
 let p_resource = take_while1 (fun c -> c != ' ') <* space
 
 let p_version =
-  string "HTTP/1." *> any_char <* crlf >>= function
+  let* v = string "HTTP/1." *> any_char <* crlf in
+  match v with
   | '1' -> return `HTTP_1_1
   | '0' -> return `HTTP_1_0
-  | v -> fail (Format.sprintf "Invalid HTTP version: %C" v)
+  | v -> failwith (Format.sprintf "Invalid HTTP version: %C" v)
 
 let header =
-  lift2
-    (fun key value -> (key, value))
-    (token <* char ':' <* ows)
-    (take_till is_cr <* crlf)
-
-let http_headers =
-  let cons x xs = x :: xs in
-  fix (fun headers ->
-      let _emp = return [] in
-      let _rec = lift2 cons header headers in
-      peek_char >>= function '\r' -> _emp | _ -> _rec)
-  >>| Http.Header.of_list
-  <* crlf
+  let+ key = token <* char ':' <* ows
+  and+ value = take_while not_cr <* crlf in
+  (key, value)
+
+let http_headers r =
+  let rec aux () =
+    match peek_char r with
+    | Some '\r' -> crlf r; []
+    | _ ->
+      let h = header r in
+      h :: aux ()
+  in
+  Http.Header.of_list (aux ())
 
 let[@warning "-3"] http_request t =
-  match end_of_input t with
+  match at_end_of_input t with
   | true -> Stdlib.raise_notrace End_of_file
   | false ->
-      let meth = p_meth t in
-      let resource = p_resource t in
-      let version = p_version t in
-      let headers = http_headers t in
-      let encoding = Http.Header.get_transfer_encoding headers in
-      commit t;
-      { Http.Request.meth; resource; version; headers; scheme = None; encoding }
+    let meth = p_meth t in
+    let resource = p_resource t in
+    let version = p_version t in
+    let headers = http_headers t in
+    let encoding = Http.Header.get_transfer_encoding headers in
+    { Http.Request.meth; resource; version; headers; scheme = None; encoding }
```

Comment:
Using the current HEAD, you should be able replace this by a call to `Http.Request.make`

Reply:
OK, I've rebased and updated to use that now.